### ISSUE TITLE
texinfo example blocks can contain macros

### ISSUE THREAD

**Issue by cbm755** [catch22/octave-doctest#102]
presumably most commonly `@var{abc} = 7`.  I think these should be expanded to `ABC = 7`.

Maybe other things are done too, people could define custom macros for example...  But at some point, parsing the texinfo ourselves is just silly.

For the record, if we just required everyone to use `>>` then we could just have `help()` do the conversion and work with the plain text, dropping all of our texinfo hackery... Just sayin' as I know no one likes `>>` but me ;-)

**Comment by oheim** [comment]
I'd say that the number of relevant macros is quite small. Using `@var` inside an example is semantically questionable (if the documentation presents a concrete example there are no variables). Converting `@var` parameters into upper case is also wrong, because this kind of formatting applies to info (plain text) files only.

However, one macro that should be supported is `@print{}` for lines that are printed to console but are not a return value of a function (e. g. warnings, disp output, …).

We should probably start a list of best practive texinfo markup in the Octave wiki and discuss the desired style with other maintainers before adding support for it.

**Comment by oheim** [comment]
You can safely ignore custom macros, because they would certainly not work with the help command (unless they are defined within the function's documentation itself). So the only relevant macros should be those from octave/etc/macros.texi (`@xbackslashchar`, others should be irrelevant) and the standard macros supplied by TexInfo.

**Comment by catch22** [comment]
Colin: Hey, I'm part of the `>>` fan club ;-)

Oliver: Doctest captures all output and errors, so from our perspective `@print{}` is not be necessary (but we can interprete it appropriately if desired by the Octave community).

**Comment by oheim** [comment]
I think that `@print{}` should be used in combination with `@result{}` where appropriate.

Here we have no return value but only output:
````
disp ('abc')
@print{} abc
````

Here it is an return value:
````
x = disp ('abc')
@result{} x = abc
````

Here we have a combination of both:
````
disp ('abc'); x = disp ('def')
@print{} abc
@result{} x = def
````

Here the warning is not part of the result:
````
inv (0)
@print{} warning: inverse: matrix singular to machine precision, rcond = 0
@result{} ans = Inf
````

Here we have two results:
````
[m, n] = size (zeros (1, 2))
@result{} m =  1
@result{} n =  2
````

Intermediate results and printing:
````
[m, n] = size (zeros (1, 2)), disp ('abc')
@result{} m =  1
@result{} n =  2
@print{} abc
````

**Comment by catch22** [comment]
What I mean is that doctest does not provide any support for explicitly distinguishing between printed outputs and return values. This is by design, because (1) its purpose is to emulate an interactive session at the Octave prompt and (2) results are prefixed by the variable names or by `ans =`, and therefore there is no real gain in the explicit encoding. For example, I can obtain a valid doctest for your above examples by essentially copy & pasting from an Octave session:

~~~matlab
% >> disp ('abc')
% abc
%
% >> x = disp ('abc')
% x = abc
%
% >> disp ('abc'); x = disp ('def')
% abc
% x = def
%
% >> inv (0)
% warning: matrix singular to machine precision
% ans = Inf
%
% >> [m, n] = size (zeros (1, 2))
% m =  1
% n =  2
%
% >> [m, n] = size (zeros (1, 2)), disp ('abc')
% m =  1
% n =  2
% abc
~~~

This is a big appeal of doctests from my perspective, so I don't want to compromise at least when using the `>>` syntax.

In the case of the Texinfo syntax, we can attempt to `@result` and `@print` appropriately and I am happy to accept pull requests for this.

**Comment by oheim** [comment]
@catch22, yes it is clear that this distinction is irrelevant to the doctest functionality. However, it is used throughout the Octave manual as a semantical distinction for the reader. Copy and pasting the whole session is a good start to create examples for the documentation, but converting it to TexInfo should add more than just `@example` and `@end example` to it.

Before reinventing the wheel by trying to parse various TexInfo macros and eventually failing we should stop at this point. Correctly parsing the macros is a pain in the neck and more trouble awaits down the lane, e. g. the user could switch document encoding, in a .texinfo file there could even be user defined macros within an `@example` block, and future TexInfo versions might bring further trouble.

The idea from @cbm755 to simply use the `help` command is the right approach. It is necessary to do a little hacking for detection of `@example` blocks and to carry over the `@comment doctest: +switch`, but that should already suffice. Then we could additionally need a little bit of result wrapping heuristics as described in #11. I can work on this topic and prepare a pull request, which adds support for arbitrary TexInfo sources.

**Comment by cbm755** [comment]
Great discussion, +1 for having octave/texinfo do the parsing.

Lots of the texinfo examples above should be tests: they could go into xfail for now.

**Comment by cbm755** [comment]
> Converting `@var` parameters into upper case is also wrong

oops you're absolutely right, for some reason I thought HTML did it too but no.

**Comment by cbm755** [comment]
IMHO `@print` should block 0.4.1 but the proper parsing need not land that quickly...

**Comment by catch22** [comment]
@oheim Right, that's what I had in mind. Let's do the right thing for Texinfo documentation without changing the basic `>>` syntax.

**Comment by oheim** [comment]
Solved by #107
